hands-on-exercise/1-native/2-build-module.md

```diff
@@ -0,0 +1,972 @@
+    }
+
+    return cdc.MustMarshalJSON(gs)
+}
+```
+
+Note how you comment out the lines that will eventually register the missing message and query servers and client.
+
+</CodeGroupItem>
+<CodeGroupItem title="depinject.go">
+
+```go [module/depinject.go]
+package module
+
+import (
+    "cosmossdk.io/core/address"
+    "cosmossdk.io/core/appmodule"
+    "cosmossdk.io/core/store"
+    "cosmossdk.io/depinject"
+
+    "github.com/cosmos/cosmos-sdk/codec"
+    authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+    modulev1 "github.com/alice/checkers/api/module/v1"
+    "github.com/alice/checkers/keeper"
+)
+
+var _ appmodule.AppModule = AppModule{}
+
+// IsOnePerModuleType implements the depinject.OnePerModuleType interface.
+func (am AppModule) IsOnePerModuleType() {}
+
+// IsAppModule implements the appmodule.AppModule interface.
+func (am AppModule) IsAppModule() {}
+
+func init() {
+    appmodule.Register(
+        &modulev1.Module{},
+        appmodule.Provide(ProvideModule),
+    )
+}
+
+type ModuleInputs struct {
+    depinject.In
+
+    Cdc          codec.Codec
+    StoreService store.KVStoreService
+    AddressCodec address.Codec
+
+    Config *modulev1.Module
+}
+
+type ModuleOutputs struct {
+    depinject.Out
+
+    Module appmodule.AppModule
+    Keeper keeper.Keeper
+}
+
+func ProvideModule(in ModuleInputs) ModuleOutputs {
+    // default to governance as authority if not provided
+    authority := authtypes.NewModuleAddress("gov")
+    if in.Config.Authority != "" {
+        authority = authtypes.NewModuleAddressOrBech32Address(in.Config.Authority)
+    }
+
+    k := keeper.NewKeeper(in.Cdc, in.AddressCodec, in.StoreService, authority.String())
+    m := NewAppModule(in.Cdc, k)
+
+    return ModuleOutputs{Module: m, Keeper: k}
+}
+```
+
+Good thing that it is possible to copy it.
+
+</CodeGroupItem>
+<CodeGroupItem title="autocli.go">
+
+```go [module/autocli.go]
+package module
+
+import (
+    autocliv1 "cosmossdk.io/api/cosmos/autocli/v1"
+)
+
+// AutoCLIOptions implements the autocli.HasAutoCLIConfig interface.
+func (am AppModule) AutoCLIOptions() *autocliv1.ModuleOptions {
+    return &autocliv1.ModuleOptions{
+        Query: nil,
+        Tx:    nil,
+    }
+}
+```
+
+At the moment there are no commands to define, so it stays simple.
+
+</CodeGroupItem>
+</CodeGroup>
+
+---
+
+Once more, you need to run:
+
+```sh
+$ go mod tidy
+```
+
+Your checkers module is not prepared as a standalone module. It is time to integrate it into `chain-minimal`.
+
+## Integrate checkers in `chain-minimal`
+
+Back to `chain-minimal`, you have to direct it to integrate your checkers module. It is known with the `github.com/alice/checkers` package, and located in a parallel folder.
+
+Direct it to the right dependency, so it does not look for it on `github.com`:
+
+```diff [go.mod]
+...
+    replace (
++      github.com/alice/checkers => ../checkers-minimal/
```

Comment:
How did you make it know where to find this package otherwise?

Alternatively, I am considering using a Git submodule in `chain-minimal`'s `x/` folder.

Reply:
The change is fine. Ehat I meant was that it's not obvious in the tutorial that the file you need to edit is `go.mod`.

Reply:
aah